### DISCUSSION TITLE
pip vs %pip

### DISCUSSION THREAD

**Discussion by barucsi** [LaunchX-InnovaccionVirtual/CursoIntroPython#60] [Q&A] [answered]
Hola, ¿Alguien sabe cuál es la diferencia entre usar pip y %pip?

Ya que cuando instalamos ipywidgets se usa solo el 'pip', pero para los paquetes de numpy y matplotlib se utiliza '%pip'.

Gracias y éxito para todos :)

**Comment by DanielGalindoRios** [comment]
Hola.  Si no me equivoco, en el live anterior, comentaron que este depende de la version de python que se tenga instalado, o desde donde se ejecute. Fer comento que hizo el curso desde una Rasp, y por eso algunos comandos son diferentes.

Espero me corrijan o aclaren.

Saludos

**Reply by barucsi** [reply]
Gracias! Si, de hecho no me jalaba cuando usaba !pip sino hasta que usé %pip

**Comment by Dfredude** [comment, marked as answer]
`%pip` y `!pip` son para poder correr el comando en una notebook de Jupyter. Y de hecho la forma correcta es con `%pip`, puedes ver que esa no la marca de color rojo. Aqui puedes ver que hacen lo mismo, pero de preferencia usa `%pip`
![image](https://user-images.githubusercontent.com/72330887/153352642-26b41727-38f8-4fdc-a55c-600b18ba0d3b.png)

Fuera de **_Jupyter notebook_** donde normalmente ese comando se usa, solo se puede correr en el CLI (cmd en windows) y ahi es donde solo se escribe `pip`.
![image](https://user-images.githubusercontent.com/72330887/153352814-23259fbf-1690-4594-a73d-aa9d9fa5ade3.png)


Espero haberte ayudado 💯

**Reply by barucsi** [reply]
Gracias! Si, de hecho el propio VSC hace la sugerencia de cambiar ! por %, lo que no tenía claro era el porque de solo usar pip, pero al parecer es para cmd

**Comment by buc-io** [comment]
Hola Baruc! te dejo info que encontré en internet espero te sirva, saludos. https://stackoverflow.com/questions/65547021/difference-between-pip-and-pip-in-python-jupyter-notebook-and-colab

**Reply by barucsi** [reply]
Gracias!

**Comment by doguedogue** [comment]
- **!PIP** - para uso en Jupyter Notebooks (visual code)
- **%PIP** - para uso en Jupyter Notebooks (visual code)
- **PIP** - para uso en la terminal (cmd, powershel, bash)

No olvides marcar como contestada si así lo consideras 😁

![image](https://user-images.githubusercontent.com/23409026/153479343-1703d5a4-3e38-456a-9534-548017961a6d.png)

**Reply by barucsi** [reply]
Gracias! Creo que también se puede usar pip en VSC, porque al instalar ipywidgets desde el nb, funcionó al usar solo el pip

**Reply by venegas55** [reply]
fijate que me pasa lo de un comentario arriba, vscode me marca error con !pip en notebook de jupyter... por qué será?

**Reply by barucsi** [reply]
![imagen](https://user-images.githubusercontent.com/99311757/153533177-91fcb9c4-0e4c-4273-994f-397d8c020e0f.png)
Si, de hecho sale esta leyenda VSC automaticamente hace la sugerencia de usar %

Aquí viene algo de información sobre eso
https://github.com/microsoft/vscode-jupyter/wiki/Installing-Python-packages-in-Jupyter-Notebooks#pip-install-vs-pip-install

**Comment by barucsi** [comment]
Hola! Ayer preguntaba sobre la diferencia entre usar pip y %pip en Jupyter y agradezco a quienes colaboraron con las respuestas y comentarios. Quiero compartirles algo que despejó un poco más mi duda: https://github.com/microsoft/vscode-jupyter/wiki/Installing-Python-packages-in-Jupyter-Notebooks#pip-install-vs-pip-install

Y mi conclusión (se aceptan correcciones😉) : Según la Wiki de Microsoft para Jupyter, se recomienda utlizar el comando `python -m pip install` para instalar los módulos en lugar de utilizar `pip install`. Esto de acuerdo a la documentación de instalación de los módulos de python.

Entonces, cuando nosotros utilizamos `!pip install <module>` Jupyter lo leerá como el comando de Shell `pip install <module>` que según la Wiki de Microsoft no es recomendado. En cambio cuando usamos `%pip install <module>` es el comando Shell que tendremos será `python -m pip install <module>` osea, el comando que nos recomienda Microsoft.

Resumiendo: `%pip` se traduce como el comando que Microsoft recomienda `python -m pip instal` y `pip` es la traducción del comando que no está recomendado.

Ambos funcionan `pip` y `%pip` solo que el hecho que lo recomiende Microsoft supongo que nos sirve para lo que comentaba la Mission  Comander Fer en la clase, que existen formas de "Estandarizar" ciertas funciones para que funcione en la mayoría de dispositivos.
